source/distributed/fully_distributed_tria_util.cc

```diff
@@ -0,0 +1,510 @@
+        auto add_vertices_of_cell_to_vertices_owned_by_loclly_owned_cells =
+          [](auto &cell, auto &vertices_owned_by_loclly_owned_cells) mutable {
+            // add vertices belonging to a periodic neighbor
+            for (unsigned int i = 0; i < GeometryInfo<dim>::faces_per_cell; i++)
+              if (cell->has_periodic_neighbor(i))
+                {
+                  auto face_t = cell->face(i);
+                  auto face_n = cell->periodic_neighbor(i)->face(
+                    cell->periodic_neighbor_face_no(i));
+                  for (unsigned int j = 0;
+                       j < GeometryInfo<dim>::vertices_per_face;
+                       j++)
+                    {
+                      vertices_owned_by_loclly_owned_cells.insert(
+                        face_t->vertex_index(j));
+                      vertices_owned_by_loclly_owned_cells.insert(
+                        face_n->vertex_index(j));
+                    }
+                }
+
+            // add local vertices
+            for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+                 v++)
+              vertices_owned_by_loclly_owned_cells.insert(
+                cell->vertex_index(v));
+          };
+
+        // check if multilevel hierarchy should be constructed
+        if (tria_pft.is_multilevel_hierarchy_constructed() == false)
+          {
+            AssertThrow(
+              tria.has_hanging_nodes() == false,
+              ExcMessage(
+                "Hanging nodes are only supported if multilevel hierarchy is constructed!"));
+
+            // 1) collect vertices of active locally owned cells
+            std::set<unsigned int> vertices_owned_by_loclly_owned_cells;
+            for (auto cell : tria.cell_iterators())
+              if (cell->active() && cell->subdomain_id() == my_rank)
+                add_vertices_of_cell_to_vertices_owned_by_loclly_owned_cells(
+                  cell, vertices_owned_by_loclly_owned_cells);
+
+            // helper function to determine if cell is locally relevant
+            // (i.e. a cell which is connected via a vertex via a locally owned
+            // active cell)
+            auto is_locally_relevant = [&](auto &cell) {
+              for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+                   v++)
+                if (vertices_owned_by_loclly_owned_cells.find(
+                      cell->vertex_index(v)) !=
+                    vertices_owned_by_loclly_owned_cells.end())
+                  return true;
+              return false;
+            };
+
+            // 2) process all local and ghost cells: setup needed data
+            // structures and collect all locally relevant vertices
+            // for second sweep
+            std::map<unsigned int, unsigned int> vertices_locally_relevant;
+            cell_infos.push_back(std::vector<CellData<dim>>());
+            auto &part = cell_infos[0];
```

Comment:
same here with the alias -- just use the full reference to the variable. (I also don't know what "part" would actually mean ;-)

Reply:
Regarding the name `part`: the terminology used to be different (and also the concept). As you will see in a next PR, there will be master process which creates `ConstructionData`s for some partitions  and send them to the slave-processes.